src/operator/control_flow.cc

```diff
@@ -363,6 +363,393 @@ ForeachGradient(const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ogra
+                                      const std::vector<OpReqType>& req,
+                                      const std::vector<NDArray>& outputs) {
+  // inputs are dl / df(x)
+  // outputs are dl / dx
+  // where f is the current function,
+  // x is the input to the current function,
+  WhileLoopState &state = state_ptr.get_state<WhileLoopState>();
+  const WhileLoopParam& params = state.params;
+  // sanity checks
+  CHECK_EQ(inputs.size(), (size_t) params.num_outputs);
+  CHECK_EQ(outputs.size() + 2U, (size_t) params.num_args);
+  CHECK_EQ(outputs.size() + params.num_out_data, inputs.size());
+  for (auto x : req) {
+    CHECK_NE(x, kWriteInplace);
+  }
+  for (auto x: outputs) {
+    CHECK_EQ(x.storage_type(), kDefaultStorage);
+  }
+  for (size_t i = 1; i < params.func_var_locs.ndim(); ++i) {
+    CHECK_LT(params.func_var_locs[i - 1], params.func_var_locs[i]);
+  }
+  // collect var_locs and out_locs, positions other than var_locs are out_locs, i.e.
+  // [0, var_locs[0])
+  // (var_locs[1], var_locs[2]),
+  // (var_locs[2], var_locs[3]),
+  // ...
+  // (var_locs[-2], var_locs[-1] = params.num_args - 2)
+  std::vector<dim_t> var_locs(params.func_var_locs.begin(), params.func_var_locs.end());
+  var_locs.push_back((dim_t) params.num_args - 2U);
+
+  // vectors for the backward loop
+  std::vector<NDArray> ograds(inputs);
+  std::vector<NDArray> igrads(outputs.size());
+  std::vector<OpReqType> iter_req(req.size());
+  for (int step = (int) state.n_iterations - 1; step >= 0; --step) {
+    // ograds[ : num_out_data] = inputs[ : num_out_data][step]
+    // ograds[num_out_data: ] is maintained in the end of each loop
+    std::transform(std::begin(inputs),
+                   std::begin(inputs) + params.num_out_data,
+                   std::begin(ograds),
+                   [step] (const NDArray &a) { return a.At(step); } );
+    // igrads[i] = 
+    //    outputs[i]            (step == 0)
+    //    outputs[i]            (step != 0 && i not in loop_var_locs)
+    //    ArrayLike(outputs[i]) (step != 0 && i in loop_var_locs)
+    // iter_req = 
+    //    kWriteTo              (step != 0           && i in loop_var_locs)
+    //    req[i]                (step == 0           && i in loop_var_locs)
+    //    kAddTo                (step != n_iters - 1 && i not in loop_var_locs)
+    //    req[i]                (step == n_iters - 1 && i not in loop_var_locs)
+    {
+      size_t i = 0;
+      for (size_t loc : var_locs) {
+        for ( ; i < loc; ++i) {
+          // locs other that var_locs
+          igrads[i] = outputs[i];
+          iter_req[i] = (step + 1 == (int) state.n_iterations)
+                      ? req[i]
+                      : kAddTo;
+        }
+        if (i < (size_t) params.num_args) {
+          // a var
+          igrads[i] = (step == 0)
+                    ? outputs[i]
+                    : NDArray(outputs[i].shape(), outputs[i].ctx(), true, outputs[i].dtype());
```

Comment:
same here. we can potentially avoid memory allocation here.

Reply:
Marked as TODO par our discussion